### PR TITLE
[Bots] Add Data Bucket support to Bot Spell Entries.

### PR DESCRIPTION
**Pull request by Aeadoin**
Functionality mimics Merchant Data Buckets https://docs.eqemu.io/server/scripting/merchant-data-buckets/ (Will write up a doc as well)
Data Bucket changes only occur on loading of the Bot (this is to reduce number of calls performed)

Adds 3 fields to bot_spells_entries "bucket_name", "bucket_value", "bucket_comparison"

I have compiled and tested existing functionality, and with use of data buckets/comparisons.

### CODE REVIEW

File: zone/bot.cpp [hunk 1]
```diff
@@ -10323,6 +10324,30 @@ void Bot::SpawnBotGroupByName(Client* c, std::string botgroup_name, uint32 leade
 	);
 }
 
+bool Bot::GetBotOwnerDataBuckets()
+{
+	auto bot_owner = GetBotOwner();
+    if (!bot_owner) {
```
Comment (Kinglykrab) [inline zone/bot.cpp:10330]: Formatting.

File: zone/botspellsai.cpp [hunk 2]
```diff
@@ -1604,6 +1604,11 @@ bool Bot::AIHealRotation(Mob* tar, bool useFastHeals) {
 std::list<BotSpell> Bot::GetBotSpellsForSpellEffect(Bot* botCaster, int spellEffect) {
 	std::list<BotSpell> result;
 
+	auto bot_owner = botCaster->GetBotOwner();
+    if (!bot_owner) {
```
Comment (Kinglykrab) [inline zone/botspellsai.cpp:1608]: Formatting.

File: zone/botspellsai.cpp [hunk 3]
```diff
@@ -1624,14 +1629,20 @@ std::list<BotSpell> Bot::GetBotSpellsForSpellEffect(Bot* botCaster, int spellEff
 			);
 				auto player_value = botCaster->bot_data_buckets[full_name];
 				if (player_value.empty()) {
-				continue;
+				full_name = fmt::format(
+					"{}-{}",
+					bot_owner->GetBucketKey(),
+					bucket_name
+				);
```
Comment (Kinglykrab) [inline zone/botspellsai.cpp:1636]: Formatting.

File: zone/botspellsai.cpp [hunk 4]
```diff
@@ -1649,6 +1660,11 @@ std::list<BotSpell> Bot::GetBotSpellsForSpellEffect(Bot* botCaster, int spellEff
 std::list<BotSpell> Bot::GetBotSpellsForSpellEffectAndTargetType(Bot* botCaster, int spellEffect, SpellTargetType targetType) {
 	std::list<BotSpell> result;
 
+	auto bot_owner = botCaster->GetBotOwner();
```
Comment (Kinglykrab) [inline zone/botspellsai.cpp:1663]: Formatting.

File: zone/botspellsai.cpp [hunk 5]
```diff
@@ -1669,14 +1685,20 @@ std::list<BotSpell> Bot::GetBotSpellsForSpellEffectAndTargetType(Bot* botCaster,
 			);
 				auto player_value = botCaster->bot_data_buckets[full_name];
 				if (player_value.empty()) {
-					continue;
+				full_name = fmt::format(
+					"{}-{}",
+					bot_owner->GetBucketKey(),
+					bucket_name
+				);
+					player_value = botCaster->bot_data_buckets[full_name];
```
Comment (Kinglykrab) [inline zone/botspellsai.cpp:1693]: Formatting.

File: zone/botspellsai.cpp [hunk 6]
```diff
@@ -1716,14 +1743,20 @@ std::list<BotSpell> Bot::GetBotSpellsBySpellType(Bot* botCaster, uint32 spellTyp
 			);
 				auto player_value = botCaster->bot_data_buckets[full_name];
 				if (player_value.empty()) {
-					continue;
```
Comment (Kinglykrab) [inline zone/botspellsai.cpp]: Formatting.

File: zone/bot.cpp (hunk 1 above)
Comment (Aeadoin) [reply zone/bot.cpp]: Fixed

File: zone/botspellsai.cpp (hunk 3 above)
Comment (Aeadoin) [reply zone/botspellsai.cpp]: Fixed

File: zone/botspellsai.cpp (hunk 4 above)
Comment (Aeadoin) [reply zone/botspellsai.cpp:1665]: Fixed

File: zone/botspellsai.cpp (hunk 6 above)
Comment (Aeadoin) [reply zone/botspellsai.cpp]: Fixed

File: zone/botspellsai.cpp (hunk 2 above)
Comment (Aeadoin) [reply zone/botspellsai.cpp]: Fixed

File: zone/botspellsai.cpp [hunk 7]
```diff
@@ -2898,15 +2899,12 @@ bool Bot::AI_AddBotSpells(uint32 iDBSpellsID) {
...
 	for (auto &e : spell_list->entries) {
-		if (
-			EQ::ValueWithin(e.minlevel, e.maxlevel, GetLevel()) &&
-			e.spellid &&
-			!IsSpellInBotList(spell_list, e.spellid)
-		) {
-			if (!CheckDataBucket(e.bucket_name, e.bucket_value, e.bucket_comparison)) {
-				continue;
+		if (EQ::ValueWithin(GetLevel(), e.minlevel, e.maxlevel) && e.spellid) {
+			if(!e.bucket_name.empty() && !e.bucket_value.empty()) {
```
Comment (Kinglykrab) [inline zone/botspellsai.cpp:2903]: if (

File: zone/botspellsai.cpp [hunk 8]
```diff
@@ -2840,14 +2840,15 @@ bool Bot::AI_AddBotSpells(uint32 iDBSpellsID) {
...
 		for (auto &e : parentlist->entries) {
 			if (
-				EQ::ValueWithin(e.minlevel, e.maxlevel, GetLevel()) &&
+				EQ::ValueWithin(GetLevel(), e.minlevel, e.maxlevel) &&
 				e.spellid &&
 				!IsSpellInBotList(spell_list, e.spellid)
 			) {
-				if (!CheckDataBucket(e.bucket_name, e.bucket_value, e.bucket_comparison)) {
-					continue;
+				if(!e.bucket_name.empty() && !e.bucket_value.empty()) {
```
Comment (Kinglykrab) [inline zone/botspellsai.cpp:2847]: if (

**Review by Kinglykrab — APPROVED** [review]

**Pull request merged**